lib/puppet/pops/types/types.rb

```diff
@@ -1,415 +1,1482 @@
+    #
+    class PAnyType < TypedModelObject
+      # Checks if _o_ is a type that is assignable to this type.
+      # If _o_ is a `Class` then it is first converted to a type.
+      # If _o_ is a Variant, then it is considered assignable when all its types are assignable
+      # @return [Boolean] `true` when _o_ is assignable to this type
+      # @api public
+      def assignable?(o)
+        case o
+          when Class
+          # Safe to call _assignable directly since a Class never is a Unit or Variant
+          _assignable?(Puppet::Pops::Types::TypeCalculator.singleton.type(o))
+        when PUnitType
+          true
+        when PVariantType
+          # Assignable if all contained types are assignable
+          o.types.all? { |vt| assignable?(vt) }
+        when PNotUndefType
+          if !(o.type.nil? || o.type.assignable?(PUndefType::DEFAULT))
+            assignable?(o.type)
+          else
+            _assignable?(o)
+          end
+        else
+          _assignable?(o)
         end
+      end
+
+      # Returns `true` if this instance is a callable that accepts the given _args_
+      #
+      # @return [Boolean]
+      def callable?(args)
+        args.is_a?(PAnyType) && kind_of_callable? && args.callable_args?(self)
+      end
+
+      # Returns `true` if this instance is considered valid as arguments to _callable_
+      # @return [Boolean]
+      # @api private
+      def callable_args?(callable)
+        false
+      end
+
+      # Generalizes value specific types. Types that are not value specific will return `self` otherwize
+      # the generalized type is returned.
+      #
+      # @return [PAnyType] The generalized type
+      # @api public
+      def generalize
+        # Applicable to all types that have no variables
+        self
+      end
+
+      # Responds `true` for all callables, variants of callables and unless _optional_ is
+      # false, all optional callables.
+      # @return [Boolean] `true`if this type is considered callable
+      # @api private
+      def kind_of_callable?(optional=true)
+        false
+      end
+
+      def hash
+        self.class.hash
+      end
+
+      # Returns true if the given argument _o_ is an instance of this type
+      # @return [Boolean]
+      def instance?(o)
+        true
+      end
+
+      def ==(o)
+        self.class == o.class
+      end
+
+      alias eql? ==
+
+      def to_s
+        Puppet::Pops::Types::TypeCalculator.string(self)
+      end
 
-        def hash
-          self.class.hash
+      # The default instance of this type. Each type in the type system has this constant
+      # declared.
+      #
+      DEFAULT = PAnyType.new
+
+      protected
+
+      # @api private
+      def _assignable?(o)
+        o.is_a?(PAnyType)
+      end
+
+      NAME_SEGMENT_SEPARATOR = '::'.freeze
+
+      # @api private
+      def class_from_string(str)
+        begin
+          str.split(NAME_SEGMENT_SEPARATOR).inject(Object) do |memo, name_segment|
```

Comment:
prefer ´reduce´ over ´inject´ (i.e. we use map/reduce instead of collect/inject)